### PR TITLE
Update comparison-restic.md

### LINKED ISSUE #10: Comparison with Restic
I just discovered this now. I'm wondering if you could create some sort of feature comparison table to show how this compares to restic (and maybe even other popular tools) so that we can easily evaluate if this will suit our needs.

Moreover, perhaps it would be worth adding some detail to the readme explaining why this exists as opposed to just continuing to contribute to restic or others tools?

### PR DESCRIPTION
**Pull request by aawsome**
closes #10

### CODE REVIEW

File: src/comparison-restic.md [hunk 1]
```diff
@@ -1,31 +1,113 @@
...
+  cache | ✅ | ❌
+  cat | ✅ | ✅
+  config | ❌ (no in-repo config) | ✅
+  check | ✅ | ✅                           
+  copy | ✅ | ✅                           
+  diff | ✅ | ✅                          
+  dump | ✅ | ✅                          
+  find | ✅ | ❌                          
+  forget | ✅ | ✅                          
+  | generate | completions
```
Comment (simonsan) [inline src/comparison-restic.md:46]:
```suggestion
shell completions | generate | completions
```

Comment (aawsome) [reply src/comparison-restic.md]: is overworked in another way

File: src/comparison-restic.md [hunk 2]
```diff
@@ -1,31 +1,379 @@
...
-- New command `webdav` to serve repository as webdav service
-- New command `merge`
-- New command `repo-info`
-- `check` command checks and uses cache; option `--trust-cache` is available
-- Option `prune --fast-repack` for faster repacking
-- Syntax `<SNAPSHOT>[:PATH]` is available for many commands
-
-## Missing points
-
-- [ ] tests and benchmarks
```
Comment (simonsan) [inline src/comparison-restic.md:30]: That thing is now completely missing. I think it's important to mention that `rustic` is lacking `testing` and currently should not be used in production. When we have decent coverage, of let's say 60% (currently it's around 3-4%), I think we can remove that note. Also, we need to get rid of the panics in `rustic_core`.

Comment (simonsan) [reply src/comparison-restic.md:30]: Also, the `benchmark` TODO I found good, so maybe opening an issue "Add benchmarks to rustic/restic comparison" is good.

Comment (aawsome) [reply src/comparison-restic.md:30]: I think we have the "lacking tests" disclaimer much more prominent on the main rustic README - here we basically talk about a feature comparison. And honestly it is hard to compare quality. That would introduce to include some measurements like code coverage, number of open and known bugs, number of people using it, github stars... - And most of these measurements we don't even know (are not tracked) for restic. Does restic have a code coverage of 60%? I don't know!

So, while I fully agree that this is something we want to work on in rustic and we want to let our users be aware of, I still think this is nothing were we can give a qualified comparison with restic.

About benchmark comparisons I opened #63.

Comment (simonsan) [reply src/comparison-restic.md:30]: Ah, no, misunderstanding. I explicitly mean that such a comparison here could lead people to believe that 'rustic' is already an alternative right now — although in my eyes it can't actually be because of the problems mentioned. People should also realize from this comparison that they are beta testers for `rustic` and should not use it in production until we have solved the problems around testing and panicking. I'm on Windows, and for example, I can't really run `prune` without `rustic` panicking these days. Meaning, I can't fully use it as a backup solution right now.

I think it's really important to not raise expectations that rustic currently is a fully functioning alternative. Rather make people understand that they help us develop an alternative in Rust for the future by using it in parallel and opening issues.

We can't currently assure people that want to use `rustic` that it is working as expected, because we can't proof that the expectation implemented in test cases and run against the code base will succeed.

Personally, at this stage, I would totally stop implementing new features (with or without new tests), and concentrate only on cleaning up and maintaining the code base to fix the technical debt as we did with the `docs`.

- `rustic` itself needs some more tests that include byte-by-byte comparison of files, to make sure it produces viable backups, maybe even some interoperability tests with `restic` running aside and comparing the outcomes

- `rustic_core` needs some more love regarding the implementation of foundations for testing, so we can test components more easily, e.g. the various backend implementations, but also the various components of the algorithms for packing, pruning, etc.

Correctness and this stage is worth more (IMHO), than new features where we don't know (or can't proof) if the foundations are doing what they should.

So currently a disclaimer like this should be sufficient, for the comparison: 

> ### **Note**: Even though it might seem that rustic can be a complete restic alternative, we have to point out that it should not be used as an alternative in production systems right now due to missing tests and general problems with panics in the code. We are currently working on fixing the issues to make sure `rustic` is the reliable backup solution you'd expect!✊

Comment (aawsome) [reply src/comparison-restic.md:30]: I partly disagree: rustic **is** an alternative to restic. And it can and is used as alternative and also in production environment (as I personally do, for instance).
I am not saying that rustic has the same maturity as restic has and there might be edge cases which have bugs, of course. Also due to the much lower user base there are higher probabilities that such bugs have not been encountered so far. 
But on the other hand, also restic has bugs in edge cases and also has panics in the code. I do agree that we should invest in more tests and I am willing to do so. Adding tests sometimes helps to find bugs and moreover helps when you refactor code. It also sometimes prevents bug from entering by future changes. But for users, basically two things do matter: Do exist bugs along the way of my personal workflow? And if, do they threaten my data in the backup? Fact is, these typical workflows IMO are already pretty stable in rustic. And there was so far no data-critical bug reported by a user.

Moreover, note that there is a fundamental difference between restic and rustic: restic is implemented in Go while rustic in Rust. As an effect there are many features which are implemented using only few lines of code whereas the corresponding restic implementation is much bigger. This indicates that more unknown bugs could be in the restic implementation, actually.

But on the end it's all speculation. Fact is that missing tests do not imply that bugs exist! And also having many tests does not imply that a software is bug-free.

Also, to achieve maturity we do need people testing and using rustic. It doesn't help telling people: "Don't use this software, we are not even trusting it outselves". I personally do trust rustic and can say that it is a great piece of software!

TL;DR: We can add some disclaimer (actually already have added it in the main README) but don't need to be too pessimistic about rustic. Also IMO no need to repeat ourselves too often.

> rustic itself needs some more tests that include byte-by-byte comparison of files, to make sure it produces viable backups, maybe even some interoperability tests with restic running aside and comparing the outcomes

rustic already *does* have tests which include a byte-by-byte comparison of files, namely [here](https://github.com/rustic-rs/rustic/blob/main/tests/backup_restore.rs)

About your prune bug: Is there already an issue for it?

Comment (simonsan) [reply src/comparison-restic.md:30]: I do think we are on the same line here, I would just wish, that our focus would completely lay on improvements to testing right now and focussing on the roadmap and maintaining and stabilizing the current features.

Before adding any new features, that someone requests via an issue. I understand that implementing features is fun. The problem right now is, that the foundations IMHO don't allow implementing features confidently. Rusts type system makes it harder to produce a certain category of bugs, logical bugs it can not prevent, though. So more features will grow the - untested - surface of rustic itself as well.

> And there was so far no data-critical bug reported by a user.

The concern is, that as you said as well, it's unclear, if it hasn't been discovered because of the smaller user base. And my stance is, before we ask people to widely test everything with their production data, we should actually assure that rustic works the way it was/is intended. If a user reports such a bug, it essentially means, we're too late with testing. Because essentially that would be something, we may have caught with better testing.

> Fact is that missing tests do not imply that bugs exist! And also having many tests does not imply that a software is bug-free.

Exactly, but the absence of tests overall is making it impossible to reason about it. There should be at least a test for every major use cases implemented, so we can assure users, that rustic does what it was intended to do. Even if that intention is wrong, but at least, then we can change the test to match expectations and then implement something to make it work.

Right now, it's the opposite. We don't really know if `rustic` is really doing what it should for the use case, then there are refactorings and features added for use cases, but we still can't reason about if it's actually doing what it should. The work, that tests would do and assure us of, we kind of load off to end user-based testing, that's why I agree, that `beta` is a good term for it. And I think we both agree that it would be reasonable, to bring `rustic` out of beta and be more confident about it doing what it should.

> "Don't use this software, we are not even trusting it outselves"

That I don't want to tell. I want to explicitly say: Don't use this for anything important. Use it in addition to your current backup procedure, don't give up on that procedure, yet. We will update you, when we feel rustic is mature enough, that you can fully switch to it.

> rustic already does have tests which include a byte-by-byte comparison of files, namely here

dircmp uses hashes: https://crates.io/crates/dircmp

> About your prune bug: Is there already an issue for it?

I would consider this issue, to at least make it better, as rustic panics due to too many connections and backoff (so it's a mixture of a lot of different concerns, connection handling, parallelism, default settings, provider settings, backoff settings, etc.) A part of the error handling could be to apply a strategy for backing off due to being connection limited and lowering the amount of connections automatically. A `check --read-data` can never run through successfully like this, though.

https://github.com/rustic-rs/rustic_core/issues/140

Comment (simonsan) [reply src/comparison-restic.md:30]: > TL;DR: We can add some disclaimer (actually already have added it in the main README) but don't need to be too pessimistic about rustic. Also IMO no need to repeat ourselves too often.

Yeah, I think it's just that this overview now implies things that I'm personally not so sure are good to imply at the current stage - WITHOUT a disclaimer that rustic is in `beta`. I wouldn't even think it is being pessimistic about rustic, it's more like being realistic (as in healthy questioning) about the capabilities and confidence in correctness that a software with not a lot of testing can have (even in Rust).

Because people will probably get to this page without a warning, directly from a search engine etc. I think it is worth it, to make people aware, also here. So people are aware. 

#### Otherwise, it raises wrong expectations, which is worse to disappoint than to have lower expectations and be happy about them being fulfilled.

**Review by simonsan — APPROVED** [review]
